internal/controllers/extension_controller.go

```diff
@@ -0,0 +1,378 @@
+		ext.Status.ResolvedBundleResource = ""
+		setResolvedStatusConditionUnknown(&ext.Status.Conditions, "validation has not been attempted as spec is invalid", ext.GetGeneration())
+
+		setDeprecationStatusesUnknown(&ext.Status.Conditions, "deprecation checks have not been attempted as spec is invalid", ext.GetGeneration())
+		return ctrl.Result{}, nil
+	}
+
+	// gather vars for resolution
+	vars, err := r.variables(ctx)
+	if err != nil {
+		ext.Status.InstalledBundleResource = ""
+		setInstalledStatusConditionUnknown(&ext.Status.Conditions, "installation has not been attempted due to failure to gather data for resolution", ext.GetGeneration())
+		ext.Status.ResolvedBundleResource = ""
+		setResolvedStatusConditionFailed(&ext.Status.Conditions, err.Error(), ext.GetGeneration())
+
+		setDeprecationStatusesUnknown(&ext.Status.Conditions, "deprecation checks have not been attempted due to failure to gather data for resolution", ext.GetGeneration())
+		return ctrl.Result{}, err
+	}
+
+	// run resolution
+	selection, err := r.Resolver.Solve(vars)
+	if err != nil {
+		ext.Status.InstalledBundleResource = ""
+		setInstalledStatusConditionUnknown(&ext.Status.Conditions, "installation has not been attempted as resolution failed", ext.GetGeneration())
+		ext.Status.ResolvedBundleResource = ""
+		setResolvedStatusConditionFailed(&ext.Status.Conditions, err.Error(), ext.GetGeneration())
+
+		setDeprecationStatusesUnknown(&ext.Status.Conditions, "deprecation checks have not been attempted as resolution failed", ext.GetGeneration())
+		return ctrl.Result{}, err
+	}
+
+	// lookup the bundle in the solution that corresponds to the
+	// ClusterExtension's desired package name.
+	bundle, err := r.bundleFromSolution(selection, ext.Spec.Source.Package.Name)
+	if err != nil {
+		ext.Status.InstalledBundleResource = ""
+		setInstalledStatusConditionUnknown(&ext.Status.Conditions, "installation has not been attempted as resolution failed", ext.GetGeneration())
+		ext.Status.ResolvedBundleResource = ""
+		setResolvedStatusConditionFailed(&ext.Status.Conditions, err.Error(), ext.GetGeneration())
+
+		setDeprecationStatusesUnknown(&ext.Status.Conditions, "deprecation checks have not been attempted as resolution failed", ext.GetGeneration())
+		return ctrl.Result{}, err
+	}
+
+	// TODO: kapp-controller integration
+	// * lookup the bundle in the solution/selection that corresponds to the Extension's desired Source
+	// * set the status of the Extension based on the respective deployed application status conditions.
+
+	// Now we can set the Resolved Condition, and the resolvedBundleSource field to the bundle.Image value.
+	ext.Status.ResolvedBundleResource = bundle.Image
+	setResolvedStatusConditionSuccess(&ext.Status.Conditions, fmt.Sprintf("resolved to %q", bundle.Image), ext.GetGeneration())
+
+	// TODO: Question - Should we set the deprecation statuses after we have successfully resolved instead of after a successful installation?
```

Comment:
This is a question I had in the past that I don't think was fully answered. I ended up leaving it after a successful installation because if the installation fails and the previous installation is _not_ deprecated we shouldn't mark the Extension as deprecated. IMO once a deprecated extension bundle is installed that is when the deprecation status should be set.

Reply:
Leaving it to be after successful installation for now.